sklearn/naive_bayes.py

```diff
@@ -1528,3 +1541,600 @@ def _joint_log_likelihood(self, X):
+            Target values.
+
+        classes : array-like of shape (n_classes,), default=None
+            List of all the classes that can possibly appear in the y vector.
+
+            Must be provided at the first call to partial_fit, can be omitted
+            in subsequent calls.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            Weights applied to individual samples (1. for unweighted).
+
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        first_call = not hasattr(self, "classes_")
+        if first_call:
+            self._validate_params()
+            self._check_feature_names(X, reset=True)
+            self._check_n_features(X, reset=True)
+            self._validate_estimators(check_partial=True)
+            self._validate_column_callables(X)
+        else:
+            self._check_feature_names(X, reset=False)
+            self._check_n_features(X, reset=False)
+        # Consistency checks for X, y are delegated to subestimators
+
+        # Subestimators get original sample_weight. This is for class counts:
+        if sample_weight is not None:
+            weights = _check_sample_weight(sample_weight, X=y, copy=True)
+
+        # Subestimators should've checked classes. We set classes_ for counts
+        # and so that first_call becomes False at next partial_fit call.
+        _check_partial_fit_first_call(self, classes)
+
+        # We don't use sklearn.utils.multiclass.class_distribution, because it
+        # neither returns class_count, nor is suitable for partial_fit.
+        if sample_weight is None:
+            counts = np.zeros(len(self.classes_))
+            for i, c in enumerate(self.classes_):
+                counts[i] = (column_or_1d(y) == c).sum()
+        else:
+            counts = np.zeros(len(self.classes_))
+            for i, c in enumerate(self.classes_):
+                counts[i] = (weights * (column_or_1d(y) == c)).sum()
+
+        if first_call:
+            self.n_classes_ = len(self.classes_)
+            self.class_count_ = counts
+        else:
+            self.class_count_ += counts
+
+        estimators = list(self._iter(fitted=not first_call, replace_strings=True))
+        fitted_estimators = Parallel(n_jobs=self.n_jobs)(
+            delayed(_partial_fit_one)(
+                estimator=clone(nb_estimator) if first_call else nb_estimator,
+                X=_safe_indexing(X, cols, axis=1),
+                y=y,
+                message_clsname="ColumnwiseNB",
+                message=self._log_message(name, idx, len(estimators)),
+                classes=classes,
+                sample_weight=sample_weight,
+            )
+            for idx, (name, nb_estimator, cols) in enumerate(estimators, 1)
+        )
+        self._update_fitted_estimators(fitted_estimators)
+        self._update_class_prior()
+        return self
+
+    @property
+    def _estimators(self):
```

Comment:
I wonder if we need those. It is not clear yet why we cannot set directly `estimators` in `get_params` similarly to `BaseHeterogeneousEnsemble`.

Reply:
We need them. `_BaseComposition` expects a list of pairs `[('name', est), ...]`, as in `_BaseHeterogeneousEnsemble`. But we have a list of triplets `[('name', est, cols), ...]`. The sole purpose of `_estimators` is to convert them to pairs. `_BaseComposition._set_params` can't handle `cols` specifically, but is good for things like `'name__alpha'`, so the adaptor from triplets to pairs is a thing worth having. I follow the implementation of `ColumnTranformer`, which deals with the same problem.